### ISSUE TITLE
Make ChangeNotifier a mixin

### ISSUE THREAD

**Issue by Levi-Lesches** [flutter/flutter#40808]
[`ChangeNotifier`](https://api.flutter.dev/flutter/foundation/ChangeNotifier-class.html) is meant to be mixed-in to custom "state" classes, and I use this for state management (please don't yell at me, yes I know what BLoC is). Problem is, `ChangeNotifier` is not declared as a mixin, so the linter will warn against using it, even though it is perfectly safe. 

## Steps to Reproduce

1. Add `prefer_mixin` to your `analysis_options.yaml` file. 
2. Make the following class: `class Temp with ChangeNotifier {}`
3. Get berated. 

## Logs

This is a static analysis issue, no `flutter run` needed. 

Relevant Dartanalyzer error:
```
INFO|LINT... Prefer using mixins.
```

Output of `flutter doctor -v`:
```
[√] Flutter (Channel stable, v1.9.1+hotfix.2, on Microsoft Windows [Version 10.0.18362.356], locale en-US)
    • Flutter version 1.9.1+hotfix.2 at C:\Users\myUser\flutter
    • Framework revision 2d2a1ffec9 (12 days ago), 2019-09-06 18:39:49 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0


[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\myUser\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.4)
    • Android Studio at C:\Program Files\Android
    • Flutter plugin version 35.3.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

**Comment by sma** [comment]
The warning is also triggered if you use the new Router "by the book" like so:
```dart
class BlockRouterDelegate extends RouterDelegate<BlockRoutePath>
    with ChangeNotifier, PopNavigatorRouterDelegateMixin<BlockRoutePath> {
   ....
}

class BlockRoutePath { }
```

I cannot extend from `ChangeNotifier` which would fix the warning, because I have to extend from `RouterDelegate` already.

**Comment by luanpotter** [comment]
Would it be acceptable to make a PR to create `ChangeNotifierMixin` and  `WidgetsBindingObserverMixin` as a migration path as per proposed [here](https://github.com/flutter/flutter/pull/77826)?

Or should the prefer_mixin rule just be removed?

**Comment by kleeb** [comment]
the same warning is present in Flutter 2.8.1

**Comment by Levi-Lesches** [comment]
Correct, this is turning out to be a bit of a fundamental issue with how mixins are treated compared to regular classes. See https://github.com/dart-lang/language/issues/1942 for an in-depth discussion on how to move forward.

**Comment by goderbauer** [comment]
I am going to close this since at this point, this would be pretty breaking. `ChangeNotifier` was implemented before dart had mixins. There have been discussions around changing the lint `prefer_mixin` to not warn on these legacy mixins that have been implemented as classes.
